test/livebook_teams/hubs/team_client_test.exs

```diff
@@ -542,13 +565,431 @@ defmodule Livebook.Hubs.TeamClientTest do
+          secrets: [livebook_proto_secret],
+          file_systems: []
+        )
+
+      {:ok, pid} = TeamClient.start_link(team)
+      assert_receive {:hub_connected, ^id}
+      refute_receive {:secret_created, ^secret}
+
+      send(pid, {:event, :agent_connected, agent_connected})
+      assert_receive {:secret_created, ^secret}
+      assert secret in TeamClient.get_secrets(team.id)
+    end
+
+    test "replaces the secret with updated value", %{agent_name: agent_name, node: node} do
+      team = build_agent_team_hub(node)
+      id = team.id
+
+      secret =
+        build(:secret,
+          name: "SECRET_UPDATED",
+          value: "an encrypted value",
+          hub_id: id
+        )
+
+      secret_key = Livebook.Teams.derive_key(team.teams_key)
+      secret_value = Livebook.Teams.encrypt(secret.value, secret_key)
+      livebook_proto_secret = LivebookProto.Secret.new!(name: secret.name, value: secret_value)
+
+      org_key = erpc_call(node, :get_org_key!, [team.org_key_id])
+      org_key_pair = erpc_call(node, :create_org_key_pair, [[org: org_key.org]])
+      deployment_group = erpc_call(node, :create_deployment_group, [[org: org_key.org]])
+
+      livebook_proto_deployment_group =
+        LivebookProto.DeploymentGroup.new!(
+          id: to_string(deployment_group.id),
+          name: deployment_group.name,
+          mode: to_string(deployment_group.mode),
+          secrets: []
+        )
+
+      agent_connected =
+        LivebookProto.AgentConnected.new!(
+          id: deployment_group.id,
+          name: agent_name,
+          public_key: org_key_pair.public_key,
+          deployment_group: livebook_proto_deployment_group,
+          secrets: [livebook_proto_secret],
+          file_systems: []
+        )
+
+      {:ok, pid} = TeamClient.start_link(team)
+      assert_receive {:hub_connected, ^id}
+      refute_receive {:secret_created, ^secret}
+
+      send(pid, {:event, :agent_connected, agent_connected})
+      assert_receive {:secret_created, ^secret}
+
+      updated_secret = %{secret | value: "an updated value"}
+      secret_value = Livebook.Teams.encrypt(updated_secret.value, secret_key)
+
+      updated_livebook_proto_secret =
+        LivebookProto.Secret.new!(name: updated_secret.name, value: secret_value)
+
+      agent_connected =
+        LivebookProto.AgentConnected.new!(
+          id: deployment_group.id,
+          name: agent_name,
+          public_key: org_key_pair.public_key,
+          deployment_group: livebook_proto_deployment_group,
+          secrets: [updated_livebook_proto_secret],
+          file_systems: []
+        )
+
+      send(pid, {:event, :agent_connected, agent_connected})
+      assert_receive {:secret_updated, ^updated_secret}
+
+      refute secret in TeamClient.get_secrets(team.id)
+      assert updated_secret in TeamClient.get_secrets(team.id)
+    end
+
+    test "replaces the secret with deployment group secret", %{agent_name: agent_name, node: node} do
```

Comment:
A large chunk of each test is the same setup, so I think we should either try to encapsulate it in private functions, or have a single test for secrets, so that update and delete tests don't require the elaborate setup :)